lib/Cryptol.cry

```diff
@@ -323,3 +324,87 @@ groupBy : {each,parts,elem} (fin each) =>
   [parts * each] elem -> [parts][each]elem
 groupBy = split`{parts=parts}
 
+/**
+ * Logical implication
+ */
+(==>) : {a} a -> a -> a
+a ==> b = not a || b
+
+/**
+ * Logical negation
+ */
+not : {a} a -> a
+not a = ~ a
+
+/**
+ * Conjunction
+ */
+and : {n} (fin n) => [n]Bit -> Bit
+and xs = ~zero == xs
+
+/**
+ * Disjunction
+ */
+or : {n} (fin n) => [n]Bit -> Bit
+or xs = zero != xs
+
+/**
+ * Conjunction after apping a predicate to all elements
+ */
+all : {a,n} (fin n) => (a -> Bit) -> [n]a -> Bit
+all f xs = and (map f xs)
+
+/**
+ * Disjunction after applying a predicate to all elements
+ */
+any : {a,n} (fin n) => (a -> Bit) -> [n]a -> Bit
+any f xs = or (map f xs)
+
+/**
+ * Map a function over an array
+ */
+map : {a, b, n} (a -> b) -> [n]a -> [n]b
+map f xs = [f x | x <- xs]
+
+/**
+ * Functional left fold.
+ *
+ * foldl (+) 0 [1,2,3] = ((0 + 1) + 2) + 3
+ */
+foldl : {a, b, n} (fin n) => (a -> b -> a) -> a -> [n]b -> a
+foldl f acc xs = ys ! 0
+ where ys = [acc] # [f a x | a <- ys | x <- xs]
+
+/**
+ * Compute the sum of the words in the array.
+ */
+sum : {a,n} (fin n, Arith a) => [n]a -> a
+sum xs = foldl (+) zero xs
+
+/**
+ * Scan left is like a fold that emits the intermediate values
+ */
+scanl : {b, a, n} (fin n) => (b -> a -> b) -> b -> [n]a -> [n+1]b
+scanl f acc xs = ys
+ where
+  ys = [acc] # [f a x | a <- ys | x <- xs]
+
+/**
+ * Zero extension.
+ */
+extend : {n,m} (fin m, fin n, m >= n) => [n] -> [m]
+extend n = zero # n
+
+/**
+ * Returns true if the predicate is satisfied by at most one element in the
+ * array.
+ */
+uniquely : {n,a} (fin n) => (a -> Bit) -> [n]a -> Bit
+uniquely p xs =  0b10 > (us ! 0)
+  where us = [0b00] # [ u || u + extend [p x] | x <- xs | u <- us ]
```

Comment:
@TomMD I was thinking of a definition that uses only boolean operations instead of mixing bitwise and arithmetic. I've used something like this before:

    oneof : {n} (fin n) => [n] -> Bit
    oneof xs = bs ! 0
      where
        as = [True] # [ a && (~ x) | a <- as | x <- xs ]
        bs = [False] # [ if x then a else b | a <- as | b <- bs | x <- xs ]